src/graphics.c

```diff
@@ -49,95 +49,46 @@ struct graphics * graphics_create( struct graphics *parent )
-			const char* arg4 = strtok(0, " ");
-			int x;
-			int y;
-			int w;
-			int h;
-            if (!arg1 || !str2int(arg1, &x) || !arg2 || !str2int(arg2, &y) || !arg3 || !str2int(arg3, &w) || !arg4 || !str2int(arg4, &h) || window == -1) {
-                return -1;
-            }
-		    graphics_clear(current->windows[window], x, y, w, h);
-        } else if (!strcmp(action, "line")) {
-			const char* arg1 = strtok(0, " ");
-			const char* arg2 = strtok(0, " ");
-			const char* arg3 = strtok(0, " ");
-			const char* arg4 = strtok(0, " ");
-			int x;
-			int y;
-			int w;
-			int h;
-            if (!arg1 || !str2int(arg1, &x) || !arg2 || !str2int(arg2, &y) || !arg3 || !str2int(arg3, &w) || !arg4 || !str2int(arg4, &h) || window == -1) {
-                return -1;
-            }
-		    graphics_line(current->windows[window], x, y, w, h);
-        } else if (!strcmp(action, "text")) {
-			const char* arg1 = strtok(0, " ");
-			const char* arg2 = strtok(0, " ");
-			const char* arg3 = strtok(0, "\"");
-            int x;
-            int y;
-            if (!arg1 || !str2int(arg1, &x) || !arg2 || !str2int(arg2, &y) || !arg3 || window == -1) {
-                return -1;
-            }
-			int i;
-			for (i = 0; arg3[i]; i++) {
-			    graphics_char(current->windows[window],	x+i*8, y, arg3[i]);
-			} 
-        } else if (!strcmp(action, "char")) {
-			const char* arg1 = strtok(0, " ");
-			const char* arg2 = strtok(0, " ");
-			const char* arg3 = strtok(0, " ");
-            int x;
-            int y;
-            if (!arg1 || !str2int(arg1, &x) || !arg2 || !str2int(arg2, &y) || !arg3 || window == -1) {
-                return -1;
-            }
-            graphics_char(current->windows[window],	x, y, arg3[0]);
-		}
-        action = strtok(0, " ");
+    char* str;
+    struct graphics_color c;
+
+    while (command && command->type) {
+        switch (command->type) {
+            case WINDOW:
+                window = command->args[0];
+                if (window < 0 || window >= current->window_count) {
+                    return -1;
+                }
+                break;
+            case COLOR:
+                c.r = command->args[0];
+                c.g = command->args[1];
+                c.b = command->args[2];
+                c.a = 0;
+                graphics_fgcolor(current->windows[window], c);
+                break;
+            case RECT:
+                graphics_rect(current->windows[window], command->args[0], command->args[1], command->args[2], command->args[3]);
+                break;
+            case CLEAR:
+                graphics_clear(current->windows[window], command->args[0], command->args[1], command->args[2], command->args[3]);
+                break;
+            case LINE:
+                graphics_line(current->windows[window], command->args[0], command->args[1], command->args[2], command->args[3]);
+                break;
+            case TEXT:
+                str = (char*)command->args[2];
```

Comment:
Make a note in here that all the graphics commands pass in arguments by value, while the string passes in by reference.  That works for now, but might cause trouble if the graphics language is e.g. passed through a pipe.